Committee-based consensus (#42)
ssvlabs/SIPs: sips/cluster_consensus.md

|     Author     |           Title            |  Category  |       Status        |    Date    |
| -------------- | -------------------------- | ---------- | ------------------- | ---------- |
| Matheus Franco | Cluster consensus          | Core       | open-for-discussion | 2024-03-05 |

## Summary

> Aggregate `Attestation` and `Sync Committee` duties based on the cluster of operators and the duties' slot. Also, merge the post-consensus partial signature messages for such duties into a single message.

Reviewer: I don't see in the SIP nothing about merging post consensus?

Author: Added more detail

<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Currently, an operator manages many `Validator` objects. Each has a `DutyRunner` for a duty type, each with its own `QBFTController` with its unique ID (that is also inserted in its associated messages).

<p align="center">
<img src="./images/cluster_consensus/previous_scheme.drawio.png"  width="50%" height="10%">
</p>

For the proposed change to take place, different `Validator` objects should use the same `QBFT Instance`. For that, we propose decoupling the `QBFTController` object from the `DutyRunner`.

<p align="center">
<img src="./images/cluster_consensus/new_scheme.drawio.png"  width="50%" height="10%">
</p>

### New IDs

Since a single `QBFT instance` will be responsible for several validator duties, its ID must not be dependent on a validator key but rather on a cluster of operators. For that, we propose changing the `MessageID` from

```mermaid
flowchart RL
	subgraph MessageID
		Domain
		ValidatorPublicKey
		Role
	end
```

to

```mermaid
flowchart RL
	subgraph MessageID
		Domain
		OperatorCluster
		Role
	end
```

### QBFT Controller

> The current `QBFTController` structure allows only one consensus instance at a time. This must change, extending the `QBFTController` into a router of messages for the different instances.

Reviewer: In my opinion we can simply drop `QbftController` from spec
It used to be there when a runner ran several instances at the same time...

We don't do this anymore.
Even if it is needed by the implementation people I don't see why spec needs this

Author: Specifically for this change, I think we should keep it to handle the several operator's QBFT instances running in parallel.

### DutyRunner & QBFT Controller

Since the `DutyRunner` will not have its specific `QBFTController`, it must have a way to start a consensus instance and receive its decided value. For that, we propose applying the observer design pattern by which the `DutyRunner` (observer) can be updated upon a `QBFT Instance` (observable) termination.

We suggest that the `DutyRunner` holds a reference to the operator's `QBFTController` to start a `QBFT Instance` and observe it.

```go
func (r *DutyRunner) execute() {
	r.QBFTController.StartConsensus(r.committee, r.duty, r)
}

func (c *QBFTController) StartConsensus(committee []types.Operator, duty types.Duty, observer Observer) {
	instance := c.StartInstance(committee, duty)
	instance.registerObserver(observer)
}

> ```

Reviewer: > Since the `DutyRunner` will not have its specific `QBFTController`

This wording is confusing since you later show that the runner has a direct reference to `QBFTController`

Also we shouldn't get into implementation details with the observer unless absolutely necessary for tests

Author: I see why this is confusing. Sorry for that.

What I meant is that the `DutyRunner` used to have a `QBFTController` of its own. Now, the operator has the `QBFTController` and the `DutyRunner` will hold a reference to it.

I tried not to add a lot of code. I thought that would help to show how the observer pattern could be used.

## Drawbacks

- Several duties will depend on the same consensus execution. Thus, its failure will imply many attestation misses. Nonetheless, this will also make the operator more scalable due to the overall reduction in exchanged messages and processing costs.

## Extra improvements

> - If multiple partial signatures contained in a merged message refer to the same root (i.e. validators in the same Ethereum committee), the signatures can be verified using batch verification.

Reviewer: this is just the reduce signatures idea?
If so does it belong here?

## Message Validation

This duties transformation requires similar changes in message validation, namely:
- Different consensus executions are tagged by the `MessageID`. This change would be propagated with no further issues. However, the `MessageID` is used to get the validator's public key and the duty's role which are used as an ID to store the consensus state. This must be changed to use the operators' committee and the duty's role, or even simply the `MessageID`.

> - Message validation limits the number of attestation duties per validator by using the validator's public key contained in the `MessageID`. This is no longer possible. A new limitation can be accomplished by checking the number of validators a cluster of operators is assigned to. If this number is less than 32 (the number of slots in an epoch), then we can limit the number of attestation duties of such cluster per epoch. The only exception would be if such a cluster is assigned to a sync committee duty (considering that we will indeed merge attestations and sync committee duties altogether in the same consensus execution).

Reviewer: you also update message validation in knowledge base
should be on a seperate branch...
Also google sheet should be versioned

Author: Sure. Will do it after all the changes here are agreed on.

## Open questions

> - What should be the maximum number of signatures a post-consensus message can contain? The trade-off here refers to reducing the number of exchanged messages versus reducing the impact of a DoS buffer attack attempt.

Reviewer: didn't understand this and the relation to the SIP

Author: In this SIP, I suggest merging the post-consensus signatures of the aggregated duties into a single message. However, I also say that the number of signatures in a message must be bounded. Here, I leave this issue to be opened for discussion.

<p align="center">
<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Under the new design we will have a `Cluster` object that will be the top level object in charge of processing consensus messages and partial signature messages for the attestation and sync committee roles.

The `Cluster` will hold a single `ConsensusRunner` and multiple `PartialSigRunners` for each Validator managed by the cluster.

For other duty roles the old design will remain.

#### Code

```go
// Cluster is a cluster of a unique set of operators that run the same validator set
type Cluster interface {
	// Initializes and starts the runner for the duties for the given slot
  	Start(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
	// ProcessMessage processes a message routed to this Cluster
	ProcessMessage(msg *types.SSVMessage)
}

type Cluster struct {
	ClusterRunners     map[spec.Slot]ClusterRunner
	Network           Network
	Beacon            BeaconNode
	OperatorID        OperatorID

> 	ClusterShares     [spec.ValidatorPubKey]ClusterShares

Reviewer 2: ```suggestion
	ClusterShares     map[spec.ValidatorPubKey]ClusterShares
```

	// highestDecidedSlot holds the highest decided duty slot and gets updated after each decided is reached
	highestDecidedSlot spec.Slot
}

// ClusterShare Partial Validator info needed for cluster duties
type ClusterShare interface {
	getSharePubKey()      []byte
	getCommittee()        []*Operator
	getQuorum()           uint64
	getPartialQuorum() 	  uint64
	getSigner()           BeaconSigner
}

// ClusterRunner manages the duty cycle for a certain slot
type ClusterRunner interface {
	// Start the duty lifecycle for the given slot. Emits a message.
    StartDuties(slot spec.Slot, attesterDuties []*types.Duty, syncCommit	teeDuties []*types.Duty) error
	// Processes cosensus message

>     ProcessConsensus(consensusMessage *qbft.Message) ru error

Reviewer 2: can you fix? there's some formatting issue

<p align="center">
<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Under the new design we will have a `Cluster` object that will be the top level object in charge of processing consensus messages and partial signature messages for the attestation and sync committee roles.

The `Cluster` will hold a single `ConsensusRunner` and multiple `PartialSigRunners` for each Validator managed by the cluster.

For other duty roles the old design will remain.

#### Code

```go
// Cluster is a cluster of a unique set of operators that run the same validator set
type Cluster interface {
	// Initializes and starts the runner for the duties for the given slot
  	Start(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
	// ProcessMessage processes a message routed to this Cluster
	ProcessMessage(msg *types.SSVMessage)
}

type Cluster struct {
	ClusterRunners     map[spec.Slot]ClusterRunner
	Network           Network
	Beacon            BeaconNode
	OperatorID        OperatorID

> 	ClusterShares     map[spec.ValidatorPubKey]ClusterShares

Reviewer 3: ```suggestion
	Shares     map[spec.ValidatorPubKey]ClusterShares
```

Again with Monte Carlo simulations using the Mainnet dataset, the number of attestation duties aggregated presented the following distribution.

<p align="center">
<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Under the new design we will have a `Cluster` object that will be the top level object in charge of processing consensus messages and partial signature messages for the attestation and sync committee roles.

The `Cluster` will hold a `ClusterRunner` object for each slot it is running a beacon duty for.

For other duty roles the old design will remain.

#### Code

```go
// Cluster is a cluster of a unique set of operators that run the same validator set
type Cluster interface {
	// Initializes and starts the runner for the duties for the given slot
  	Start(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
	// ProcessMessage processes a message routed to this Cluster
	ProcessMessage(msg *types.SSVMessage)
}

type Cluster struct {

> 	ClusterRunners     map[spec.Slot]ClusterRunner

Reviewer: don't see the issue
in the key space you can have multiple slots

	Network           Network
	Beacon            BeaconNode
	OperatorID        OperatorID
	ClusterShares     [spec.ValidatorPubKey]ClusterShares
}

// ClusterShare Partial Validator info needed for cluster duties
type ClusterShare interface {
	getSharePubKey()      []byte
	getCommittee()        []*Operator
	getQuorum()           uint64
	getPartialQuorum() 	  uint64
	getSigner()           BeaconSigner
	// highestAttestingSlot holds the highest slot for which attester duty ran for a validator
	highestAttestingSlot spec.Slot
}

// ClusterRunner manages the duty cycle for a certain slot
type ClusterRunner interface {
	// Start the duty lifecycle for the given slot. Emits a message.

>     StartDuties(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error

Reviewer: Cluster `Start` calls this

	// Processes cosensus message
    ProcessConsensus(consensusMessage *qbft.Message) error
	// Processes a post-consensus message
	ProcessPostConsensus(msg ClusterSignaturesMessage) 
}

type ClusterRunner struct {

> 	Share          *types.Share

Reviewer 2: is single share enough? don't we need share per validator?

        BlockRoot: vote.BlockRoot,
        Source: vote.Source,
        Target: vote.Target,
        CommiteeIndex: duty.CommitteeIndex,
        AggregationBits: bits,
    }
}
```

#### Happy Flow

1. `Cluster` receives duties that match a certain slot. If the slot is higher then `highestDecidedSlot` starts consensus for the relevant roles, and initializes a `ClusterRunner` for the relevant Validators.
2. `Cluster` receives consensus messages and hands them over `ClusterRunner` that hands them over to the `QBFTController` that has unchanged logic. The only difference is `BeaconVote` is used as the ConsensusData object.
3.  Once `ProcessConsensus` decides, the `Cluster` will create a post consensus of PartialSignatureMessage that aggregates the beacon partial signature for all relevant validators.
4. `Cluster` will process post-consensus partial signature messages and submit a beacon message for each validator.

### Stopping Runs

Previously we have letted new validator duties stop the run for the previous duty. For a cluster this is not a good idea and instead we will count on a tuned `CutOffRound` per duty to stop the instance.

### Omitting Partial Signatures

> If in post-consensus stage for attestation duty, the duty's slot is lower than a validator's `ClusterShare's` `highestAttestingSlot`, the `ClusterRunner` will omit the post-consesnus message for this validator. 

Reviewer 2: IMO this description is lacking what happens regards a quorum, and/or who created the message. if we got 2 partials and we created the third should we still omit it? if we have 2 and received a third should we omit it?

Reviewer: This section only talks about omitting messages because of slots.

#### Sync Committee
`CutOffRound = 4  \\ one slot`

#### Attestation
`CutOffRound = 12 \\ one epoch`

> ### ClusterID
> 
> An identifier for cluster must be added to `MessageID`.

Reviewer 2: ```suggestion
### MsgID + ClusterID
Today a MsgID is constructed from Validator and Duty type, in order to route messages into the cluster runners, identifier for cluster must be added. `ClusterID`.
```

```go
[48]byte ClusterID

// Return a 48 bytes ID for the cluster of operators
func getClusterID(operatorIDs []OperatorID) ClusterID {

> 	// Create a 16 bytes constant prefix for cluters

Reviewer 2: ```suggestion
	// Create a 16 bytes constant prefix for clusters
```

## P2P

### Network Topology

Previously, to communicate on behalf of a certain validator, the operators computed the topic ID in the following way:

```go
topicID := hexToUint64(validatorPKHex[:10]) % subnetsCount
```

Now, the messages are related to a committee (not to a specific validator). Thus, if we were to use the committee's validators' associated topics, we would be sending multiple equal messages on the network. To avoid that, we will start to use a topic associated with the committee. It can be computed in a way similar to the following:

```go
// sort
sort.Slice(committee, func(i, j int) bool {
	return committee[i] < committee[j]
})
// Convert to bytes
bytes := make([]byte, len(committee)*4)
for i, v := range committee {
	binary.BigEndian.PutUint32(bytes[i*4:], uint32(v))
}
// Hash

> hashed := sha256.Sum256(bytes)

Reviewer 2: isn't it the same? he is iterating over commitee

	Shares          map[ValidatorPubkey]Share
    QBFTParams     QBFTParams
	QBFTController *qbft.Controller
	BeaconNetwork  *types.BeaconNetwork
}

// BeaconVote is the consensus data
type BeaconVote struct {
    BlockRoot         phase0.Root
    Source            phase0.Checkpoint
    Target            phase0.Checkpoint
}

func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
    bits := bitfield.New(duty.CommitteeLength)
    bits.Set(duty.ValidatorCommitteeIndex)

    return Attestation{
        Slot: duty.Slot,
        BlockRoot: vote.BlockRoot,
        Source: vote.Source,
        Target: vote.Target,
        CommiteeIndex: duty.CommitteeIndex,
        AggregationBits: bits,
    }
}

func ConstructSyncCommittee(vote BeaconVote, duty AttesterDuty) SyncCommitteeMessage{
    return SyncCommitteeMessage
    {
        Slot: duty.Slot,
        BlockRoot: vote.BlockRoot,
        ValidatorIndex: duty.ValidatorIndex,
    }
}
```
#### PartialSignatureMessages

The current structure that we have in code can be unchanged.

```go
type PartialSignatureMessages struct {
	Type     PartialSigMsgType
	Slot     phase0.Slot
	Messages []*PartialSignatureMessage

> }

Author: So, indeed, the share's signatures will be merged into the same message, right? If so, what will be the maximum number of signatures per message?

Reviewer: Max validators an operator can hold times 2 I believe

// PartialSignatureMessage is a msg for partial Beacon chain related signatures
type PartialSignatureMessage struct {
	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
    Signer           OperatorID // TODO: there is an optimization where we don't have to duplicate this field
}
```

We must have the following flow:
1. When you `calculateExpectedRootsAndDomain` (similar to `sync_committee_aggregator`). Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`

> 3. When processing the messages find all the roots that have quorums, mark them, and sumbit corresponding beacon data to beacon chain.

Author: So, for attestation duties, we can relate the validators to the signing root. But what about the signatures for the sync committee duties? The signing root should be the same if I'm not mistaken.

Reviewer: It isn't the same because we sign Beacon Objects

Reviewer: https://github.com/bloxapp/ssv-spec/blob/ebb60143f17ea25edf09edeb398559d764b0e01b/ssv/sync_committee.go#L138-L143

Reviewer: you sign the head root for sync committee
for attestation it is the head root + source + target

  - If the same `ValidatorIndex` appears more than 2 times in `SignedPartialSignatureMessage.Message.Messages`, the message is rejected.

Regarding duties' specific rules,
  - The attestation's and sync committee's specific rules are dropped in favor of a common set of rules for the new duty type attestation + sync committee. Namely, the higher attestation limits are used (34 slots, 12 rounds).
  - We can no longer limit a single validator to two attestation attempts per epoch. If we were to count only attestation, we could limit a cluster with $V$ validators to do $2 \times V$ consensus per epoch. However, the fact the sync committee duties use the same consensus instances as attestations can force us to tolerate 32 consensus instances from a cluster in an epoch. Thus, we suggest limiting by $2 \times V$ executions only with a condition check that no validators of such cluster are doing sync committee duties in the epoch.
  - For the `BNRoleAttesterOrSyncCommittee` role, the number of signatures in a `PartialSignatureMessages` is limited to $min(2 * V, V+SYNC_COMMITTEE_SIZE.)$.

Regarding implementation,

>   - The `ConsensusState` is currently mapped by a `ConsensusID` that uses the validator public key and the role from the `MessageID`. Since `MessageID` will have a ClusterID, instead of a validator public key, for attestations and sync committees, the mapping of `ConsensusState` will need to change. We suggest either changing `ConsensusID` to encompass also a `ClusterID` or simply mapping `ConsensusState` by `MessageID`.

Reviewer: Message Validation section is fine by me..
But this specific section needs to be deleted before the SIP is merged

Author: Do you think this part regarding implementation should be removed from the SIP?
For me, no problem

Reviewer: only "Regarding Implementation"

Again with Monte Carlo simulations using the Mainnet dataset, the number of attestation duties aggregated presented the following distribution.

<p align="center">
<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Under the new design we will have a `Cluster` object that will be the top level object in charge of processing consensus messages and partial signature messages for the attestation and sync committee roles.

The `Cluster` will hold a `ClusterRunner` object for each slot it is running a beacon duty for.

For other duty roles the old design will remain.

#### Code

```go
// ClusterDuty implements Duty
type ClusterDuty struct {
	Slot         spec.Slot
	BeaconDuties []*BeaconDuty
}

// Cluster is a cluster of a unique set of operators that run the same validator set
type Cluster interface {
	// Initializes and starts the runner for the duties for the given slot
  	Start(duty Duty) error
	// ProcessMessage processes a message routed to this Cluster
	ProcessMessage(msg *signedSSVMessage *types.SignedSSVMessage)
}

type Cluster struct {
	Runners     map[spec.Slot]ClusterRunner

>     Shares      map[ValidatorPubkey]Share

Reviewer 4: @GalRogozinski align please

Reviewer: ok

	Network           Network
	Beacon            BeaconNode
	OperatorID        OperatorID
	QBFTParams        QBFTParams
    
}

type QBFTParams interface {
    Operators()        []OperatorID
    Quorum()           uint64
    PartialQuorum() 	  uint64
}

> =

Reviewer 4: ```suggestion

```

type Share struct {
    // New field:
	// HighestAttestingSlot holds the highest slot for which attester duty ran for a validator
    HighestAttestingSlot spec.Slot
}

// ClusterRunner manages the duty cycle for a certain slot
type ClusterRunner interface {
	// Start the duty lifecycle for the given slot. Emits a message.
    StartDuty(duty Duty) error
	// Processes cosensus message
    ProcessConsensus(consensusMessage *qbft.Message) error
	// Processes a post-consensus message
    ProcessPostConsensus(msg PartialSignatureMessages) 
}

type ClusterRunner struct {
	Shares          map[ValidatorPubkey]Share

>     QBFTParams     QBFTParams

Reviewer 4: align please

}
```
#### PartialSignatureMessages

The current structure that we have in code can be unchanged.

```go
type PartialSignatureMessages struct {
	Type     PartialSigMsgType
	Slot     phase0.Slot
	Messages []*PartialSignatureMessage
}

// PartialSignatureMessage is a msg for partial Beacon chain related signatures
type PartialSignatureMessage struct {
	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
    Signer           OperatorID // TODO: there is an optimization where we don't have to duplicate this field
}
```

We must have the following flow:
1. When you `calculateExpectedRootsAndDomain` (similar to `sync_committee_aggregator`). Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`
3. When processing the messages find all the roots that have quorums for a certain validator, mark them, and sumbit corresponding beacon data to beacon chain.
4. For the next message received attempt to complete quorum for other roots.

#### Happy Flow

> 1. `Cluster` receives duties that match a certain slot. If the slot is higher then `highestDecidedSlot` starts consensus for the relevant roles, and initializes a `ClusterRunner` for the relevant Validators.

Reviewer 4: @GalRogozinski I think we dont use highestDecidedSlot anymore, we should start consensus and then not broadcasting partial sig for a validator in case the slot is not higher than HighestAttestingSlot, correct?

Again with Monte Carlo simulations using the Mainnet dataset, the number of attestation duties aggregated presented the following distribution.

<p align="center">
<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
</p>

## Spec changes

### Design

Under the new design we will have a `Committee` object that will be the top level object in charge of processing consensus messages and partial signature messages for the attestation and sync committee roles.

The `Committee` will hold a `CommitteeRunner` object for each slot it is running a beacon duty for.

For other duty roles the old design will remain.

#### Code

```go
// CommitteeDuty implements Duty
type CommitteeDuty struct {
	Slot         spec.Slot
	BeaconDuties []*BeaconDuty
}

// Committee is a committee of a unique set of operators that have shared validators
type Committee interface {
	// Initializes and starts the runner for the duties for the given slot

>   	StartDuty(duty Duty) error

Reviewer 4: @GalRogozinski should be CommitteeDuty, correct?

Reviewer: `Duty` is an interface
But perhaps in the SIP it is indeed best to use `CommitteeDuty` for clarity

I will change

	// ProcessMessage processes a message routed to this Committee
	ProcessMessage(msg *signedSSVMessage *types.SignedSSVMessage)
}

type Committee struct {
	Runners                 map[spec.Slot]*CommitteeRunner
	Operator                types.Operator
	SignatureVerifier       types.SignatureVerifier
	CreateRunnerFn          func() *CommitteeRunner
	// Save the highest slot a validator attested to
	HighestAttestingSlotMap map[types.ValidatorPK]spec.Slot
}

// StartDuty starts a new duty for the given slot
func (c Committee) StartDuty(duty types.CommitteeDuty) error {

> 	// do we need slot?

Reviewer 4: @GalRogozinski resolve comment

### Message Validation

This duties transformation requires a propagation of changes in message validation.

Regarding syntax, the rules can stay the same.

Regarding semantics,
  - If `SSVMessage.MsgID` contains a CommitteeID (i.e. the role is attestation + sync committee) and if the CommitteeID doesn't exist in the current network, it should ignore the message.
  - If a `ValidatorIndex` in `SignedPartialSignatureMessage.Message.Messages` is incorrect, considering the ValidatorPublicKey or the CommitteeID in the MessageID, it should ignore the message.

Regarding duties' general rules,
  - The duplicated partial signature rule (same validator, slot and type but different signing root or signature) now applies only to the Proposal, Aggregate, Sync Committee Contribution, Validator Registration and Voluntary Exit duties (since attestation and sync committee partial signature messages may include the same signing root.)

>   - If the same `ValidatorIndex` appears more than 2 times in `SignedPartialSignatureMessage.Message.Messages`, the message is rejected.

Reviewer: @MatheusFranco99 
> since attestation and sync committee partial signature messages may include the same signing root.

you sure about this?

> If the same `ValidatorIndex` appears more than 2 times in `SignedPartialSignatureMessage.Message.Messages`, the message is rejected.

if we have attestation and sync committee will this hold?
I think maybe this part is due to an older version of the SIP

Author: Yes, if I'm not mistaken, two validators may be signing the same AttestationData (
```go
type AttestationData struct {
	Slot            Slot
	Index           CommitteeIndex
	BeaconBlockRoot Root
	Source          *Checkpoint
	Target          *Checkpoint
}
```
) if they belong to the same committee and agree on the vote (with the same `DomainAttester` domain.

Regarding the "2 times", the idea is that at most a validator index is both doing the attestation and sync committee duties. So, it can have at most 2 signatures related to it.

Regarding duties' specific rules,
  - The attestation's and sync committee's specific rules are dropped in favor of a common set of rules for the new duty type attestation + sync committee. Namely, the higher attestation limits are used (34 slots, 12 rounds).

>   - We can no longer limit a single validator to two attestation attempts per epoch. If we were to count only attestation, we could limit a committee with $V$ validators to do $2 \times V$ consensus per epoch. However, the fact the sync committee duties use the same consensus instances as attestations can force us to tolerate 32 consensus instances from a committee in an epoch. Thus, we suggest limiting by $2 \times V$ executions only with a condition check that no validators of such committee are doing sync committee duties in the epoch.

Reviewer 4: @MatheusFranco99 Why we allow 2 *v consensuses per epoch?
Does that mean that If a committee has 500 validators we allow up to 1000 consensuses?

Author: @lior-blox the limit would be min(32, 2*v). So for any $v>=16$, the maximum limit will always be 32 consensuses per epoch

Reviewer: This can happen more likely because of a delay of duty processing rather than a fork

>   - For the `BNRoleAttesterOrSyncCommittee` role, the number of signatures in a `PartialSignatureMessages` is limited to $min(2 * V, V+SYNC\_COMMITTEE\_SIZE.)$.

Reviewer: ```suggestion
  - For the `RunnerRole` role, the number of signatures in a `PartialSignatureMessages` is limited to $min(2 * V, V+SYNC\_COMMITTEE\_SIZE.)$.
```

### GossipSub Scoring

The only two groups of the GossipSub scoring parameters that depend on the expected message rate are $P_2$ and $P_3$. $P_3$ is currently switched off, so no change needs to be done on that end. $P_2$ regards increasing a peer score for its "first delivery" messages and counts with three parameters:
 - A decay $d_2$
 - A cap value ($cap_2$) for the counter
 - A weight ($w_2$) to multiply the counter value and sum the result to the topic's score.

The decay $d_2$ can stay the same (0.3162277660168379, which decays 1 to 0.01 in 4 epochs). The cap value is computed using the mesh target size, $D = 8$, and the expected message rate per decay interval, and it's defined as the convergence value if one keeps sending twice the number of messages it's expected to send. In other words,

$$cap_2 = \frac{2m}{D} \times \frac{1}{(1-d_2)}$$

> The message rate for a validator per decay interval (one epoch), used to be calculated as $600/10000 \times (32 \times 12)$. To account for all validators in a topic, we multiply this value by $\frac{V}{32}$. Then, to account for the improvement of this change, considering that the message rate will drop to 15% of the current value, we multiply it by $0.15$ and, finally, we get

Reviewer: In your simulation didn't you say it will drop to 21.61%?

Author: New Mainnet estimation.

> $$cap_2 = \frac{2}{D} \times \left( \frac{600}{10000} \times (32 \times 12) \times \frac{V}{128} \times 0.15 \right) \times \frac{1}{1-d_2}$$

Reviewer: In the description you still write $\frac{V}{32}$...
So actually before we had a mistake for not multiplying by this value?

Author: The description should also have V/128. Updated it
